### ISSUE TITLE
`CtModel#getAllTypes` returns a class which is not part of the AST

### ISSUE THREAD

**Issue by msteinbeck** [INRIA/spoon#3010]
Steps to reproduce:

1. Clone this repository: https://github.com/lets-blade/blade
2. Check out this revision: 926913f3eaa2bc44b4bb8703e7d61e773cccc4d0
3. Run the following code snippet (replace `/path/to/repo`):
```java
Launcher launcher = new Launcher();
launcher.addInputResource("/path/to/repo/src/main/java");
List<CtType> types = launcher.buildModel().getAllTypes().stream()
        .filter(t -> !t.getPosition().isValidPosition())
	.collect(Collectors.toList());
System.out.println(types);
```

Result:
```bash
[class HttpVersion {}]
```

If I use grep to search for `HttpVersion`:

```bash
grep -r HttpVersion *
```

I get the following output:

```bash
src/main/java/com/blade/server/netty/HttpServerHandler.java:import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
src/main/java/com/blade/server/netty/StaticFileHandler.java:import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
src/main/java/com/blade/server/netty/StaticFileHandler.java:        HttpResponse httpResponse = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
src/main/java/com/blade/server/netty/StaticFileHandler.java:        FullHttpResponse httpResponse = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, OK, staticInputStream.asByteBuf());
src/main/java/com/blade/server/netty/RouteMethodHandler.java:import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
src/main/java/com/blade/mvc/wrapper/OutputStreamWrapper.java:            HttpResponse httpResponse = new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
src/main/java/com/blade/mvc/handler/DefaultExceptionHandler.java:import io.netty.handler.codec.http.HttpVersion;
src/main/java/com/blade/mvc/handler/DefaultExceptionHandler.java:            FullHttpResponse fullHttpResponse = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.valueOf(response.statusCode()), buffer);
```

So `HttpVersion` is not part of my input source. I tested version 7.4.0 and 7.5.0-beta-21.

**Comment by nharrand** [comment]
Hello @msteinbeck ,

Thanks for the report. 
From what I can read [here](https://github.com/INRIA/spoon/blob/9db1d7754f14743f79632d8073e45c109d7aed22/src/main/java/spoon/support/compiler/jdt/ContextBuilder.java#L297), in noClasspath mode, when importing a static field, we create a class holding the field to be imported. So it seems to be intentional. But I get that it is weird that these types are listed byt `getAllTypes()`
I am not sure what behavior would be less misleading for spoon users. We could mark these types and filter them out in getAllTypes() but then it might lead to some other weird discrepancies.

WDYT @monperrus ?

**Comment by monperrus** [comment]
Indeed, there is an ambiguity in the spec: "returns all top-level types of the model".

Two options to resolve this ambiguity:

* returns all top-level types of the model, incl. shadow types (those built from binary classes)
* returns all top-level types of the model, only build from sources, excl shadow types(those built
from binary classes)

I prefer the latter, and we can also add a method "getAllTypesInclShadowTypes()"

**Comment by msteinbeck** [comment]
I think excluding shadow types (those built from binary classes) will break incremental builds, because previously compiled files are added to the classpath while creating a new increment. In the example stated above, a class which neither is created from a .class file nor has been added to the input source is returned by `getAllTypes`.

**Comment by monperrus** [comment]
OK.

Would it work if we change the incremental builder to call `getAllTypesInclShadowTypes` instead of `getAllTypes` (new version)?

**Comment by monperrus** [comment]
@msteinbeck WDYT?
